File: sdk-humanlayer/session.py
"""Session management for HumanLayer SDK."""
import uuid
import logging
from datetime import datetime, timezone
from typing import Optional, List

logger = logging.getLogger("humanlayer.session")

_current_session: Optional["Session"] = None


def get_current_session() -> Optional["Session"]:
    return _current_session


def set_current_session(session: Optional["Session"]) -> None:
    global _current_session
    _current_session = session


class Session:
    def __init__(self, name: str = "default", metadata: dict = None):
        self.session_id = str(uuid.uuid4())
        self.name = name
        self.metadata = metadata or {}
        self.start_time: Optional[datetime] = None
        self.end_time: Optional[datetime] = None
        self.status = "inactive"
        self.events: List[dict] = []
        self._client = None

        # Stats
        self.total_tokens = 0
        self.prompt_tokens = 0
        self.completion_tokens = 0
        self.llm_calls = 0
        self.tool_calls = 0
        self.errors = 0

    def bind_client(self, client) -> None:
        self._client = client

    def start(self) -> "Session":
        self.start_time = datetime.now(timezone.utc)
        self.status = "active"
        if self._client:
            self._client.send_session(self._to_dict())
        return self

    def end(self, status: str = "completed") -> None:
        self.end_time = datetime.now(timezone.utc)
        self.status = status
        if self._client:
            self._client.send_session(self._to_dict())

    def log_event(self, event: dict) -> None:
        """Called by callback handler to record an event."""
        event["session_id"] = self.session_id
        self.events.append(event)

        # Update token stats
        data = event.get("data", {})
        usage = data.get("metadata", {}).get("token_usage", {})
        if usage:
            self.total_tokens += usage.get("total_tokens", 0)
            self.prompt_tokens += usage.get("prompt_tokens", 0)
            self.completion_tokens += usage.get("completion_tokens", 0)

        etype = event.get("event_type", "")
        if "llm" in etype:
            self.llm_calls += 1
        elif "tool" in etype:
            self.tool_calls += 1
        elif "error" in etype:
            self.errors += 1

        # Enqueue to client
        if self._client:
            self._client.enqueue_event(event)

    def _to_dict(self) -> dict:
        return {
            "session_id": self.session_id,
            "name": self.name,
            "status": self.status,
            "event_count": len(self.events),
            "start_time": self.start_time.isoformat() if self.start_time else None,
            "end_time": self.end_time.isoformat() if self.end_time else None,
            "metadata": self.metadata,
            "statistics": {
                "total_tokens": self.total_tokens,
                "prompt_tokens": self.prompt_tokens,
                "completion_tokens": self.completion_tokens,
                "llm_calls": self.llm_calls,
                "tool_calls": self.tool_calls,
                "errors": self.errors,
            },
        }

    def __enter__(self):
        return self

    def __exit__(self, exc_type, *_):
        self.end(status="failed" if exc_type else "completed")

File: sdk-js/src/session.ts
import { v4 as uuidv4 } from "uuid";
import { HumanLayerClient } from "./client";
import { SessionError } from "./exceptions";

export type SessionStatus = "inactive" | "active" | "completed" | "failed";

export class Session {
  sessionId: string;
  name: string;
  metadata: Record<string, unknown>;
  status: SessionStatus = "inactive";
  startTime?: Date;
  endTime?: Date;

  // Stats
  totalTokens = 0;
  promptTokens = 0;
  completionTokens = 0;
  llmCalls = 0;
  toolCalls = 0;
  errors = 0;

  private client?: HumanLayerClient;

  constructor(name: string, metadata: Record<string, unknown> = {}) {
    this.sessionId = uuidv4();
    this.name = name;
    this.metadata = metadata;
  }

  bindClient(client: HumanLayerClient): void {
    this.client = client;
  }

  async start(): Promise<void> {
    if (this.status !== "inactive") {
      throw new SessionError(`Session already started (status: ${this.status})`);
    }
    this.status = "active";
    this.startTime = new Date();
    await this._sync();
  }

  async end(status: "completed" | "failed" = "completed"): Promise<void> {
    this.status = status;
    this.endTime = new Date();
    await this._sync();
    await this.client?.flush();
  }

  logEvent(event: object): void {
    if (!this.client) return;

    // Inject session_id into event
    const enriched = {
      ...(event as Record<string, unknown>),
      session_id: this.sessionId,
    };

    // Update stats based on event type
    const e = enriched as Record<string, unknown>;
    const eventType = e.event_type as string | undefined;
    if (eventType === "llm_end" || eventType === "llm_start") {
      if (eventType === "llm_end") {
        this.llmCalls++;
        const data = e.data as Record<string, unknown> | undefined;
        const metadata = data?.metadata as Record<string, unknown> | undefined;
        const usage = metadata?.token_usage as Record<string, unknown> | undefined;
        if (usage) {
          this.totalTokens += (usage.total_tokens as number) ?? 0;
          this.promptTokens += (usage.prompt_tokens as number) ?? 0;
          this.completionTokens += (usage.completion_tokens as number) ?? 0;
        }
      }
    } else if (eventType === "tool_end") {
      this.toolCalls++;
    } else if (
      eventType === "tool_error" ||
      eventType === "llm_error" ||
      eventType === "chain_error"
    ) {
      this.errors++;
    }

    this.client.enqueueEvent(enriched);
  }

  private async _sync(): Promise<void> {
    if (!this.client) return;
    await this.client.sendSession({
      session_id: this.sessionId,
      name: this.name,
      status: this.status,
      start_time: this.startTime?.toISOString(),
      end_time: this.endTime?.toISOString(),
      metadata: this.metadata,
      statistics: {
        total_tokens: this.totalTokens,
        prompt_tokens: this.promptTokens,
        completion_tokens: this.completionTokens,
        llm_calls: this.llmCalls,
        tool_calls: this.toolCalls,
        errors: this.errors,
      },
    });
  }
}

// Module-level current session
let _currentSession: Session | undefined;

export function getCurrentSession(): Session | undefined {
  return _currentSession;
}

export function setCurrentSession(session: Session | undefined): void {
  _currentSession = session;
}
